src/libawkward/array/ByteMaskedArray.cpp

```diff
@@ -1297,9 +1297,29 @@ namespace awkward {
       Index64 nextcarry = pair.first;
       Index64 outindex = pair.second;
 
+      if ( slicestarts.length() != length() or slicestops.length() != length() ) {
+        throw std::runtime_error("Slices look weird");
+      }
+      Index64 carrystarts(slicestarts.length() - numnull);
+      Index64 carrystops(slicestops.length() - numnull);
+      int64_t k=0;
+      for(int64_t i=0; i < length(); ++i) {
+        int64_t from = outindex.getitem_at_nowrap(i);
+        // here we could OR this with starts[i]==stops[i] to pass singletons
+        // as if they were nulls. We would have to hack a bit on nextcarry and outindex though
+        if ( from >= 0 ) {
+          carrystarts.setitem_at_nowrap(k, slicestarts.getitem_at_nowrap(i));
+          carrystops.setitem_at_nowrap(k, slicestops.getitem_at_nowrap(i));
+          k++;
+        }
+      }
+      if ( k != carrystarts.length() ) {
+        throw std::runtime_error("failed to carry starts and stops properly");
+      }
```

Comment:
Note that when an `Index` (such as `outindex`, `slicestarts`, and `slicestops`) are passed into a kernel, you have to pass the raw pointer (`.ptr().get()`) and its offset (`.offset()`). Both pieces of information are needed to get at logical index `i` of an array, since one `Index` may be sharing a buffer with another `Index`, starting at a different offset: logical index `i` is at `offset + i`. It's a downside of forcing everything through a C interface. If I had been more organized at the beginning, I might have defined some C structs to carry both pieces of information over, though it would still be non-standard syntax to call a function that adds the offset in every get/put.

Except for one special case, _output_ `Indexes` can be assumed to have offset = 0 (because you just allocated it before entering the kernel; of course the offset is 0!). The special case involves concatenation.

Reply:
Why not pass the shifted pointer into the function instead?

Reply:
That... is a great idea. Maybe a future refactoring will do that. It could be a method on `Index` to give the actual beginning of the data; it has to reference-count the base pointer, but there can be special methods to give something convenient to the C functions...